In [2]:
import pandas as pd
import numpy as np

In [3]:
# Activity 3 Solution

# Step 1: Load the dataset
df = pd.read_csv("../data/handling_anomalies_activity.csv")


In [4]:
# Step 2: Inspect the dataset
print("\nFIRST 5 ROWS")
print(df.head())

print("\nDESCRIPTIVE STATISTICS")
print(df["Transaction_Value"].describe())


FIRST 5 ROWS
   Record_ID  Transaction_Value
0          1             116.24
1          2              93.88
2          3              94.72
3          4              89.27
4          5             108.65

DESCRIPTIVE STATISTICS
count     48.000000
mean     100.190625
std       32.330544
min       10.000000
25%       91.467500
50%       97.415000
75%      108.740000
max      220.000000
Name: Transaction_Value, dtype: float64


In [5]:
# Part A - Z-Score Method


mean = df["Transaction_Value"].mean()
std = df["Transaction_Value"].std(ddof=1)

print("\nZ-SCORE CALCULATION")
print(f"Mean = {mean:.2f}")
print(f"Standard Deviation = {std:.2f}")


Z-SCORE CALCULATION
Mean = 100.19
Standard Deviation = 32.33


In [6]:
# Calculate Z-Scores
df["Z_Score"] = (df["Transaction_Value"] - mean) / std


In [7]:
# Threshold = |3|
z_outliers = df[np.abs(df["Z_Score"]) > 3]

print("\n Z-SCORE OUTLIERS")

if len(z_outliers) > 0:
    print(z_outliers)
else:
    print("No outliers detected.")


 Z-SCORE OUTLIERS
    Record_ID  Transaction_Value   Z_Score
44         45              200.0  3.087154
45         46              220.0  3.705764


In [8]:
# Part B - Modified Z-Score Method


median = df["Transaction_Value"].median()
mad = np.median(np.abs(df["Transaction_Value"] - median))

print("\n MODIFIED Z-SCORE  ")
print(f"Median = {median:.2f}")
print(f"MAD = {mad:.2f}")


 MODIFIED Z-SCORE  
Median = 97.41
MAD = 8.28


In [9]:
# Calculate Modified Z-Score
df["Modified_Z"] = 0.6745 * (
    (df["Transaction_Value"] - median) / mad
)

In [10]:
# Threshold = |3.5|
modified_outliers = df[np.abs(df["Modified_Z"]) > 3.5]

print("\n MODIFIED Z-SCORE OUTLIERS")

if len(modified_outliers) > 0:
    print(modified_outliers)
else:
    print("No outliers detected.")


 MODIFIED Z-SCORE OUTLIERS
    Record_ID  Transaction_Value   Z_Score  Modified_Z
44         45              200.0  3.087154    8.356713
45         46              220.0  3.705764    9.985940
46         47               15.0 -2.634989   -6.713637
47         48               10.0 -2.789641   -7.120944


In [11]:
# Compare Results
# 

print("\nCOMPARISON")

print("Outliers detected by Z-Score:")
print(z_outliers["Transaction_Value"].tolist())

print("\nOutliers detected by Modified Z-Score:")
print(modified_outliers["Transaction_Value"].tolist())


COMPARISON
Outliers detected by Z-Score:
[200.0, 220.0]

Outliers detected by Modified Z-Score:
[200.0, 220.0, 15.0, 10.0]


In [12]:
# Summary

print("\n SUMMARY ")

print(f"Number of Z-Score outliers: {len(z_outliers)}")
print(f"Number of Modified Z-Score outliers: {len(modified_outliers)}")


 SUMMARY 
Number of Z-Score outliers: 2
Number of Modified Z-Score outliers: 4


## INTERPRETATION 
### The Z-Score method measures how many standard deviations
### each observation is from the mean.

The Modified Z-Score method uses the median and Median Absolute Deviation (MAD), making it less sensitive to extreme values.

For this dataset, the Modified Z-Score is generally more reliable because a few extreme transaction values (such as 10, 15, 200 and 220) influence the mean and standard deviation used by the standard Z-Score.

Since the majority of transaction values are concentrated around 90–110, the median provides a better measure of central tendency, allowing the Modified Z-Score to detect anomalies more robustly.